### PR TITLE
Api middleware

### PR DESCRIPTION
**Pull request by Ofir-Smuha**
## API Middleware
- Fetched projects from mimic into projects
- Fetched data fro mimic into statistics

### CODE REVIEW

File: proj/src/components/statistics/statistics.js [hunk 1]
```diff
@@ -24,8 +27,12 @@ export default class Statistics extends Component {
     ]
   }
 
+  componentDidMount() {
+    this.props.fetchData()
+  }
+
   drawBars(){
-    let data = this.state.data || [];
+    let data = this.props.data || [];
```
Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: i don't think you actually need these next 2 lines. data can be used from props and dataKeys can be initiated with the flow statement

File: proj/src/components/statistics/statistics.js [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
   height: 100vh;
   display: flex;
   justify-content: center;
   align-items: center;
 `
 
-export default class Statistics extends Component {
+class Statistics extends Component {
 
   state = {
```
Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: let's start cleaning up component state. we have redux and mimic now so we need to remove all the junk

File: proj/src/components/statistics/statistics.js [hunk 3]
```diff
@@ -57,3 +64,9 @@ export default class Statistics extends Component {
     )
   }
 }
+
+const mapStateToProps = state => ({
+  data: state.statistics.data
```
Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: data is not the best name for variables

File: proj/src/middleware/projectsMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import axios from 'axios'
+
+const projectsMiddleware = ({dispatch}) => next => action => {
+  if (!action.meta || action.meta.type !== 'api') {
+    console.log('going to next')
```
Comment (ronlanger) [inline proj/src/middleware/projectsMiddleware.js]: remove console.log for PR's

File: proj/src/middleware/projectsMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import axios from 'axios'
+
+const projectsMiddleware = ({dispatch}) => next => action => {
+  if (!action.meta || action.meta.type !== 'api') {
```
Comment (ronlanger) [inline proj/src/middleware/projectsMiddleware.js]: wouldn't a check for meta.api be enough here?

File: proj/src/middleware/projectsMiddleware.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    console.log('going to next')
+    return next(action);
+  }
+  
+  axios.get(action.meta.url)
```
Comment (ronlanger) [inline proj/src/middleware/projectsMiddleware.js]: what if I want a post? or delete?

File: proj/src/middleware/projectsMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    .then(({data}) => dispatch(action.payload.onSuccess(data)))
+    .catch( err => console.log('error: ', err))
```
Comment (ronlanger) [inline proj/src/middleware/projectsMiddleware.js]: where is the handler for errors? this middleware is not complete yet

File: proj/src/middleware/projectsMiddleware.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import axios from 'axios'
+
+const projectsMiddleware = ({dispatch}) => next => action => {
+  if (!action.meta || action.meta.type !== 'api') {
+    console.log('going to next')
+    return next(action);
+  }
+  
+  axios.get(action.meta.url)
+    .then(({data}) => dispatch(action.payload.onSuccess(data)))
```
Comment (ronlanger) [inline proj/src/middleware/projectsMiddleware.js]: make the success and error handlers able to receive arrays of actions (onSuccess and onError can one or many actions)

File: proj/src/reducers/statistics/statisticsReducer.js [hunk 9]
```diff
@@ -1,5 +1,11 @@
 import { handleActions, combineActions } from 'redux-actions';
+import { set } from 'lodash/fp'
+const initialState = {
+  data: []
```
Comment (ronlanger) [inline proj/src/reducers/statistics/statisticsReducer.js]: too much 'data' here. like I said, not the best name. try to choose a different data structure here. what if you just made this entire reducer return an array

File: proj/src/store.js [hunk 10]
```diff
@@ -1,19 +1,18 @@
 import { createStore, compose, applyMiddleware} from 'redux'
-import thunk from 'redux-thunk';
-import persistState from 'redux-localstorage'
+// import persistState from 'redux-localstorage'
 
 import rootReducer from 'reducers/root'
-
-const middleWare = [thunk]
+import projectsMiddleware from 'middleware/projectsMiddleware';
```
Comment (ronlanger) [inline proj/src/store.js]: projectsMiddleware? isn't it apiMiddleware? please go over all the changes before submitting a PR. seems like you were in a hurry :(

File: proj/src/views/Projects.js [hunk 11]
```diff
@@ -118,4 +121,4 @@ const mapStateToProps = ({ projects }) => ({
     isLoading: projects.isLoading
 })
 
-export default connect(mapStateToProps, { openEditProject })(Projects)
\ No newline at end of file
+export default connect(mapStateToProps, { openEditProject, fetchProjects })(Projects)
```
Comment (ronlanger) [inline proj/src/views/Projects.js]: fix formatting

File: proj/src/actions/projectsActions.js [hunk 12]
```diff
@@ -26,4 +26,21 @@ export const isLoading = () => ({
...
+
+export const fetchProjects = () => ({
+  type: AT.API_REQUEST,
+  payload: {
+    onSuccess: setProjects
+  },
+  meta: {
+      type: 'api',
+      url: 'http://ofir.com/api/projects',
+      //method: post / put / detele
```
Comment (ronlanger) [inline proj/src/actions/projectsActions.js]: remove comments

File: proj/package.json [hunk 13]
```diff
@@ -3,6 +3,8 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "axios": "^0.18.0",
+    "firebase": "^5.5.5",
```
Comment (ronlanger) [inline proj/package.json:7]: not supposed to be here for this PR. commit changes that are only related to this PR

**Review by ronlanger — CHANGES_REQUESTED** [review]

**Comment by ronlanger** [comment]
There is some stuff missing. Make a habit of going through all the files/changes before submitting a PR to see that everything is in order. the reducers need structure. calling state.projects.projects is not fun.
Talk to me if any clarification is needed

File: proj/src/middleware/projectsMiddleware.js (hunk 6 above)
Comment (Ofir-Smuha) [reply proj/src/middleware/projectsMiddleware.js]: This task includes fetch only (get), the next one is with firebase with a full REST api.

File: proj/src/middleware/projectsMiddleware.js (hunk 5 above)
Comment (Ofir-Smuha) [reply proj/src/middleware/projectsMiddleware.js]: When a action fires without a meta property it cannot find "type" of undefined

Comment (Ofir-Smuha) [reply proj/src/middleware/projectsMiddleware.js]: an action*

**Review by Ofir-Smuha — COMMENTED** [review]
I fixed the issues, The only think is that there's no handling for error's yet.  so i just placed the logic without any real actions behind them.

File: proj/src/actions/statisticsActions.js [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import * as AT from 'actions/types'
+
+export const fetchData = () => ({
```
Comment (ronlanger) [inline proj/src/actions/statisticsActions.js]: on the same principle fetchData is to generic. data of what? statistics. so maybe fetchStatistics would be more appropriate

File: proj/src/actions/statisticsActions.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    onError: setError
+  },
+  meta: {
+    type: 'api',
+    url: 'http://ofir.com/api/data'
+  }
+})
+
+export const setData = (data) => ({
+  type: AT.SET_DATA,
```
Comment (ronlanger) [inline proj/src/actions/statisticsActions.js]: same here

File: proj/src/actions/types.js [hunk 16]
```diff
@@ -11,4 +11,13 @@ export const OPEN_EDIT_PROJECT = 'OPEN_EDIT_PROJECT';
...
 export const SUBMIT_PROJECT = 'SUBMIT_PROJECT';
 export const IS_LOADING = 'IS_LOADING';
-export const EDIT_PROJECT = 'EDIT_PROJECT';
\ No newline at end of file
+export const EDIT_PROJECT = 'EDIT_PROJECT';
+export const FETCH_PROJECTS = 'FETCH_PROJECTS'
+export const SET_PROJECTS = 'SET_PROJECTS'
+
+// Statistics
+export const FETCH_DATA = 'FETCH_DATA'
```
Comment (ronlanger) [inline proj/src/actions/types.js]: same here

File: proj/src/components/statistics/statistics.js [hunk 17]
```diff
@@ -1,31 +1,28 @@
...
-      {date: 'day 2', Vision: 2, RunTime: 4, Fox: 3},
-      {date: 'day 3', Vision: 3, RunTime: 8, Fox: 4},
-      {date: 'day 4', Vision: 4, RunTime: 5, Fox: 5, Nort: 2},
-      {date: 'day 5', Vision: 11, RunTime: 1, Fox: 6, Barc: 2},
-      {date: 'day 6', Vision: 6, RunTime: 4, Fox: 7},
-      {date: 'day 7', Vision: 7, RunTime: 3, Fox: 10, Nasa: 8}
-    ]
+class Statistics extends Component {
+
+  state = {}
```
Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: why leave an empty state?

File: proj/src/components/statistics/statistics.js [hunk 18]
```diff
@@ -1,31 +1,28 @@
...
+
+  componentDidMount() {
+    this.props.fetchData()
   }
 
   drawBars(){
-    let data = this.state.data || [];
+
     let dataKeys = []
```
Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: const dataKeys = flow([
      map(keys),
      flatten,
      uniq,
      filter((key) => key !== 'date')
    ])(this.props.chartData);

Comment (ronlanger) [inline proj/src/components/statistics/statistics.js]: remove the let it is not needed

File: proj/src/middleware/apiMiddleware.js [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+  axios.request({url, method})
+    .then(({data}) => {
+      if(Array.isArray(onSuccess)) {
+        onSuccess.forEach(action => dispatch(action(data)))
+      } else {
+        dispatch(action.payload.onSuccess(data))
+      }
+      })
+      .catch( err => dispatch(onError(err)));
```
Comment (ronlanger) [inline proj/src/middleware/apiMiddleware.js]: should be the same concept of onSuccess. no reason to make the on success array-ready and the onError not

File: proj/src/reducers/projects/projectsReducer.js [hunk 20]
```diff
@@ -7,11 +7,7 @@ const initialState = {
     displayEditModal: false,
     displayMode: 'list',
     currProject: null,
-    projects: {
-        1: {id: 1, name: 'Project 1'},
-        2: {id: 2, name: 'Project 2'},
-        3: {id: 3, name: 'Project 3'},
-    }
+    projects: {}
```
Comment (ronlanger) [inline proj/src/reducers/projects/projectsReducer.js]: this means you still need to access 'state.projects.projects' in connected components. you can just remove the inner 'projects' property altogether. you will also have to fix the appropriate places in the reducer to set directly in the state and not in state.projects

File: proj/src/reducers/statistics/statisticsReducer.js [hunk 21]
```diff
@@ -1,5 +1,9 @@
-import { handleActions, combineActions } from 'redux-actions';
+import { handleActions } from 'redux-actions';
 
-const initialState = {};
+const initialState = [];
 
-export default handleActions({}, initialState);
\ No newline at end of file
+export default handleActions({
+  SET_DATA: (state, {data}) => {
```
Comment (ronlanger) [inline proj/src/reducers/statistics/statisticsReducer.js]: still called data

File: proj/src/views/Projects.js [hunk 22]
```diff
@@ -118,4 +121,9 @@ const mapStateToProps = ({ projects }) => ({
     isLoading: projects.isLoading
 })
 
-export default connect(mapStateToProps, { openEditProject })(Projects)
\ No newline at end of file
+export default connect(
```
Comment (ronlanger) [inline proj/src/views/Projects.js]:
```suggestion
export default connect(mapStateToProps, {
  actionA,
  actionB
})(ComponentName)
```

File: proj/src/views/Projects.js [hunk 23]
```diff
@@ -2,13 +2,12 @@ import React, { Component } from 'react';
...
 
 import EditProject from 'components/projects/EditProject';
 import withLoader from 'components/projects/withLoader'
 import ProjectList from 'components/projects/ProjectList'
 import ProjectPreview from 'components/projects/ProjectPreview';
-import { openEditProject } from 'actions/projectsActions'
+import { openEditProject, fetchProjects } from 'actions/projectsActions'
 
 import gridLayout from 'assets/images/grid.png'
 import listLayout from 'assets/images/list.png'
```
Comment (ronlanger) [inline proj/src/views/Projects.js:13]: btw why import these images and not put directly in the src attribute?

**Review by ronlanger — CHANGES_REQUESTED** [review]

File: proj/src/views/Projects.js (hunk 23 above)
Comment (Ofir-Smuha) [reply proj/src/views/Projects.js:13]: Just looks cleaner

**Pull request merged**